src/theory/theory_model_builder.h

```diff
@@ -210,6 +215,75 @@ class TheoryEngineModelBuilder : public ModelBuilder
    */
   std::map<Node, Node> d_constantReps;
 
+  /** Theory engine model builder assigner class
+   *
+   * This manages the assignment of values to terms of a given type.
+   * In particular, it is a wrapper around a type enumerator that is restricted
+   * by a set of values that it cannot generate, called an "assignment exclusion
+   * set".
+   */
+  class Assigner
+  {
+   public:
+    Assigner() : d_te(nullptr), d_isActive(false) {}
+    /**
+     * Initialize this assigner to generate values of type tn, with properties
+     * tep and assignment exclusion set aes.
+     */
+    void initialize(TypeNode tn,
+                    TypeEnumeratorProperties* tep,
+                    const std::vector<Node>& aes);
```

Comment:
The issue is that this class is being used as (non-pointer) data in a range of a map, so its construction is done by accessing the map via:
`      Assigner& a = eqcToAssigner[eqc];`
My understand of C++ is that the default constructor will be run for data in range of maps and hence we'd end up with an extra invocation of a constructor if we overwrote the data in the map by the non-default version e.g.:
`      Assigner& a = Assigner( ... );  eqcToAssigner[eqc] = a;`
where notice `eqcToAssigner[eqc] ` constructs a 2nd object.  Or perhaps my understanding is not correct.

This is generally the reason why I introduce "initialize" methods, I do agree constructors when possible is cleaner.

Reply:
Yes, `eqcToAssigner[eqc]` will create an instance using the default constructor if `eqc` is not already in the map. However, if you know that something is already in the map, you can use [`map::at()`](https://en.cppreference.com/w/cpp/container/map/at) and to add a new element, you can use [`map::emplace()`](https://en.cppreference.com/w/cpp/container/map/emplace). If you use these methods, the class shouldn't need a default constructor.

Reply:
Emplace is not a part of my vocabulary :)  

Generally I have a strong preference for keeping the set of primitive C++ language features I use to a minimum. While the `at`/`emplace` or similar feature might do the trick, they make the code less readable in my opinion.

However, generally I agree with you that we should use constructors instead of initialize methods. In this case however, I prefer keeping this code as is since it is a single-use class that is private to TheoryModelBuilder.

Reply:
The `emplace*` methods in the STL are actually quite useful because they create objects in place (so no need for a move/copy), so they tend to be more efficient. Instead of using `at()`, using `find()` would also do the trick. Since the class is local, I am ok with leaving it as-is, though.